pkg/controller/aerospikecluster/aerospikecluster_controller.go

```diff
@@ -131,91 +137,246 @@ func (r *ReconcileAerospikeCluster) Reconcile(request reconcile.Request) (reconc
+	}
+
+	if isNew {
+		logger.Debug("It's new cluster, create empty status object")
+		if err := r.createStatus(aeroCluster); err != nil {
 			return reconcile.Result{}, err
 		}
-		return r.createCluster(aeroCluster)
+	} else {
+		logger.Debug("It's not a new cluster, check if it is failed and needs recovery")
+		hasFailed, err := r.hasClusterFailed(aeroCluster)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("Error determining if cluster has failed: %v", err)
+		}
+
+		if hasFailed {
+			return r.recoverFailedCreate(aeroCluster)
+		}
+	}
+
+	if err := r.ReconcileRacks(aeroCluster); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Setup access control.
+	if err := r.reconcileAccessControl(aeroCluster); err != nil {
+		logger.Error("Failed to reconcile access control", log.Ctx{"err": err})
+		return reconcile.Result{}, err
+	}
+
+	// Update the AerospikeCluster status.
+	if err := r.updateStatus(aeroCluster); err != nil {
+		logger.Error("Failed to update AerospikeCluster status", log.Ctx{"err": err})
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// ReconcileRacks reconcile all racks
+func (r *ReconcileAerospikeCluster) ReconcileRacks(aeroCluster *aerospikev1alpha1.AerospikeCluster) error {
+	logger := pkglog.New(log.Ctx{"AerospikeCluster": utils.ClusterNamespacedName(aeroCluster)})
+	logger.Info("Reconciling rack for AerospikeCluster")
+
+	var scaledDownRackSTSList []appsv1.StatefulSet
+	var scaledDownRackList []RackState
+
+	rackStateList := getNewRackStateList(aeroCluster)
+	for _, state := range rackStateList {
+		found := &appsv1.StatefulSet{}
+		stsName := getNamespacedNameForStatefulSet(aeroCluster, state.Rack.ID)
+		if err := r.client.Get(context.TODO(), stsName, found); err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+			// Create statefulset for rack
+			if err := r.createSTSForCluster(aeroCluster, state); err != nil {
+				return err
+			}
+			continue
+		}
+
+		// Get list of scaled down racks
+		if *found.Spec.Replicas > int32(state.Size) {
+			scaledDownRackSTSList = append(scaledDownRackSTSList, *found)
+			scaledDownRackList = append(scaledDownRackList, state)
+		} else {
+			// Reconcile other statefulset
+			if err := r.reconcileClusterSTS(aeroCluster, found, state); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Reconcile scaledDownRacks after all other racks are reconciled
+	for idx, state := range scaledDownRackList {
+		if err := r.reconcileClusterSTS(aeroCluster, &scaledDownRackSTSList[idx], state); err != nil {
+			return err
+		}
+	}
+
+	if len(aeroCluster.Status.RackConfig.Racks) != 0 {
+		// remove removed racks
+		if err := r.deleteSTSForRemovedRacks(aeroCluster, rackStateList); err != nil {
+			logger.Error("Failed to remove statefulset for removed racks", log.Ctx{"err": err})
+			return err
+		}
+	}
+
+	// Pod termination in above call will take some time. Should we wait here for sometime or
```

Comment:
See if this can be done without a sleep

Reply:
This issue, shouldn't come. only situation may be deleting a statefulset and quickly recreate with same name. Dont have a solid way to quickly check if all pods are terminated or when to stop checking. Noting it down for now.